### PR TITLE
[ONNX] Add training mode support for BatchNormalization op

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**
This commit extends the OnnxToTorch lowering for BatchNormalization op for supporting the case when training=True.

Signed-Off By: Vivek Khandelwal <vivekkhandelwal1424@gmail.com>

### CODE REVIEW

**Comment by vivekkhandelwal1** [comment]
Fixes https://github.com/nod-ai/SHARK-Turbine/issues/285

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp [hunk 1]
```diff
@@ -339,43 +339,123 @@ void mlir::torch::onnx_c::populateDefaultDomainAtoF(
...
+
+        // When training_mode=False, the op outputs only Y, where
+        // Y = (X - input_mean) / sqrt(input_var + epsilon) * scale +
+        // B
+        if (!training) {
+          rewriter.replaceOpWithNewOp<Torch::AtenBatchNormOp>(
+              binder.op, resultType, input, weight, bias, inputMean, inputVar,
+              /*training=*/cstFalse, cstMomentum, cstEps,
+              /*cudnn_enabled=*/cstFalse);
+          return success();
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:375]: If this op has a training variable, why not implement the logic below for the pytorch op and just lower with `/*training=*/training`?

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:375]: Because the PyTorch op returns only one result, while the OnnxOp expects 3 results.

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainAtoF.cpp:375]: And, the computation for `currentMean/Var` is anyway required for computing other results, hence no extra work is done here.

**Review by zjgarvey — CHANGES_REQUESTED** [review]
This does not pass e2e testing for the following onnx node tests due to a significant numerics mismatch:

"test_batchnorm_epsilon_training_mode"
"test_batchnorm_example_training_mode"

Please try to debug the numerics mismatching when you get the chance. You can either use this branch of torch-mlir in your iree-build and run the iree_tests, or you can use the alt_e2eshark and run `python run.py --torchtolinalg -t test_batch`.

**Comment by zjgarvey** [comment]
From a glance, it looks like the numerics for `Y` and `running_var` are off, but `running_mean` seems to match pretty well.

**Comment by vivekkhandelwal1** [comment]
> This does not pass e2e testing for the following onnx node tests due to a significant numerics mismatch:
> 
> "test_batchnorm_epsilon_training_mode" "test_batchnorm_example_training_mode"
> 
> Please try to debug the numerics mismatching when you get the chance. You can either use this branch of torch-mlir in your iree-build and run the iree_tests, or you can use the alt_e2eshark and run `python run.py --torchtolinalg -t test_batch`.

Hi @zjgarvey, the accuracy issue is fixed. After a lot of debugging, it turned out that the `unbiased` has to be set to False, instead of True.

**Comment by vivekkhandelwal1** [comment]
@zjgarvey Can you please review this PR, today?

**Review by zjgarvey — APPROVED** [review]
As long as this is passing numerics, the implementation is clear and well-commented on. Thanks, Vivek.
